docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,620 @@
+
+#### "As-is" startup flow
+
+The following diagram shows the "as-is" startup flow.
+The proposal would eliminate secrets-setup,
+combine kong-migrations and the kong service,
+and move edgex-proxy into the install flow.
+
+![AS-IS](./bootstrap-as-is.jpg)
+
+#### "To-be" startup flow
+
+The following diagram shows the "to-be" startup flow.
+Note that the install flow is always processed,
+but it can be short-circuited.
+
+![TO-BE](./bootstrap-to-be.jpg)
+
+#### install
+
+This is a new service to run the install flow.
+The purpose of the install flow is to generate EdgeX bootstrapping secrets
+defined in
+[Summary of bootstrapping secrets](#Summary%20of%20bootstrapping%20secrets).
+This service would be the first to run,
+and all subsequent services would implicitly or
+explicitly depend on it.
+
+This container must contain standalone executables
+for Vault, Consul, Kong, and PostgreSQL
+to be used during bootstrapping
+since these services must be started,
+stopped, and later restarted.
+Also these services should temporarily be bound
+to "localhost" to avoid exposing their services
+to the greater Docker network
+during intialization.
+For an alternative, see
+[Separate compose scripts for installation and runtime](#separate-compose-scripts-for-initialization-and-runtime).
+
+The rough flow of events is as follows:
+
+Vault bootstrapping and secrets seeding:
+
+1. Skip Vault bootstrapping if Vault's data files show evidence of completing this step.
+1. Initialize a blank Vault and immediately unseal it
+1. Encrypt the Vault keyshares and revoke the root token
+1. Perform basic initialization such as setting up the key-value store
+1. Generate needed seed secrets such as:
+   * Serf encryption key (needed for Consul)
+   * PostgreSQL superuser password (needed for Kong)
+   * Kong admin credential
+   * Redis database password
+
+Consul bootstrapping:
+
+1. Skip Consul bootstrapping if Consul's data files show evidence of completing this step.
+1. Write the production Consul configuration to a Docker volume
+   * Enable HTTP on 8500
+   * Disable DNS
+1. Start Consul for the first time and provide the Serf encryption key
+1. Generate a management token
+1. Configure the Consul secrets engine in Vault
+1. Shut down Consul
+
+Kong bootstrapping:
+
+1. Skip this step if sentinel file exists
+1. Start a temporary instance of PostgreSQL and set the superuser password
+1. Seed the Kong database schema
+1. Configure Vault for automatic credential rotation
+1. Start up a temporary instance of Kong to perform initialization
+   * Secure the Kong admin API
+   * Set up routes to internal microservices
+   * Install BYO external TLS certifiate, if provided
+1. Shut down Kong and PostgreSQL
+1. Write a sentinel file in the database storage volume
+   to indicate this install stage has completed
+
+The above process replaces the edgex-proxy microservice.
+
+Redis bootstrapping:
+
+Ths stage is still in flux,
+but should follow the same general flow above.
```

Comment:
Need to elaborate in detail more: today's secure mode Redis bootsrapping depends on `vault-worker` being ready due to the secret-store vault access token.  Not clear on how this can be done in pre-seeded fashion.

Reply:
@jim-wang-intel Please look at the updated guidance.

Reply:
@jim-wang-intel I have updated the guidance a second time.  Upon reflection, the previous guidance was not in alignment with the secret management ADR as it kept the Redis password plaintext on disk.  The updated guidance calls for creating a custom entrypoint script that dynamically generates the redis.conf to a tmpfs volume with the Redis password in it.